onnxruntime/test/framework/inference_session_test.cc

```diff
@@ -1346,66 +1348,102 @@ TEST(InferenceSessionTests, TestCopyToFromDevices) {
 
+  ASSERT_TRUE(session_object.Load(model_uri).IsOK());
   ASSERT_TRUE(session_object.Initialize().IsOK());
 }
 
-TEST(InferenceSessionTests, TestCustomTransformers) {
+// Tests a rewrite rule is applied only when the provider assigned to the node
+// is present in the compatible provider list.
+TEST(InferenceSessionTests, TestCompatibleProviders) {
   string model_uri = "testdata/transform/fusion/fuse-conv-bn-mul-add-unsqueeze.onnx";
 
   SessionOptions so;
   so.session_logid = "InferenceSessionTests.TestL1AndL2Transformers";
   so.graph_optimization_level = TransformerLevel::Level2;
-  InferenceSession session_object{so, &DefaultLoggingManager()};
-  session_object.AddCustomTransformerList({"EliminateIdentity", "ConvAddFusion", "EliminateUnsqueeze"});
-  ASSERT_TRUE(session_object.Load(model_uri).IsOK());
 
-  std::shared_ptr<Model> p_model;
-  ASSERT_TRUE(Model::Load(model_uri, p_model).IsOK());
+  InferenceSession session_object{so, &DefaultLoggingManager()};
 
-  ASSERT_TRUE(session_object.Initialize().IsOK());
-}
+  // Create rule based transformer with a dummy rewrite rule and register it with Cuda as compatible provider
+  auto dummy_rule = std::make_unique<DummyRewriteRule>("DummyRule");
+  const auto* dummy_rule_ptr = dummy_rule.get();
 
-TEST(InferenceSessionTests, DisableAllTransformers) {  
-  string model_uri = "testdata/transform/fusion/fuse-conv-bn-mul-add-unsqueeze.onnx";
+  auto graph_transformer = std::make_unique<TopDownRuleBasedTransformer>("CUDATopDownTransformer",
+                                                                         "Registered for CUDA");
+  graph_transformer->Register(std::move(dummy_rule));
+  session_object.RegisterGraphTransformer(std::move(graph_transformer), {onnxruntime::kCudaExecutionProvider});
 
-  SessionOptions so;
-  so.session_logid = "InferenceSessionTests.DisableAllTransformers";
-  so.graph_optimization_level = TransformerLevel::Default;
-  InferenceSession session_object{so, &DefaultLoggingManager()};
-  ASSERT_TRUE(session_object.Load(model_uri).IsOK());
+  // Create rule based transformer with a dummy rewrite rule and register it with CPU as compatible provider
+  auto dummy_rule1 = std::make_unique<DummyRewriteRule>("DummyRule1");
+  const auto* dummy_rule1_ptr = dummy_rule1.get();
 
-  std::shared_ptr<Model> p_model;
-  ASSERT_TRUE(Model::Load(model_uri, p_model).IsOK());
+  auto graph_transformer1 = std::make_unique<TopDownRuleBasedTransformer>("CPUTopDownTransformer", "Registered for CPU");
+  graph_transformer1->Register(std::move(dummy_rule1));
+  session_object.RegisterGraphTransformer(std::move(graph_transformer1), {onnxruntime::kCpuExecutionProvider});
 
+  session_object.Load(model_uri);
   ASSERT_TRUE(session_object.Initialize().IsOK());
+
+  // Validate transformer registered with CUDA as compatible provider is not called.
+  ASSERT_FALSE(dummy_rule_ptr->IsRewriteRuleInvoked());
+
```

Comment:
Is this really testing that TopDownRuleBasedTransformer checks the providers list? If so it seems like a test for that class rather than one for InferenceSession

Reply:
yeah I can move this test out of this file...